src/prom_agg.rs

```diff
@@ -0,0 +1,156 @@
+        lowest_time.into(),
+        greatest_time.into(),
+        step_size,
+        range,
+        time.into(),
+        val,
+        fc,
+    )
+    .internal()
+}
+pub fn prom_delta_transition_inner(
+    state: Option<Inner<GapfillDeltaTransition>>,
+    lowest_time: pg_sys::TimestampTz,
+    greatest_time: pg_sys::TimestampTz,
+    step_size: Milliseconds, // `prev_now - step` is where the next window starts
+    range: Milliseconds,     // the size of a window to delta over
+    time: pg_sys::TimestampTz,
+    val: f64,
+    fc: pg_sys::FunctionCallInfo,
+) -> Option<Inner<GapfillDeltaTransition>> {
+    unsafe {
+        in_aggregate_context(fc, || {
+            if time < lowest_time || time > greatest_time {
+                error!("input time less than lowest time")
+            }
+
+            let mut state = state.unwrap_or_else(|| {
+                let state: Inner<_> = GapfillDeltaTransition::new(
+                    lowest_time,
+                    greatest_time,
+                    range,
+                    step_size,
+                    false,
+                    false,
+                )
+                .into();
+                state
+            });
+
+            state.add_data_point(time, val);
+
+            Some(state)
+        })
+    }
+}
+
+#[pg_extern()]
+pub fn prom_delta_final(state: Internal) -> Option<Vec<Option<f64>>> {
+    prom_delta_final_inner(unsafe { state.to_inner() })
+}
+pub fn prom_delta_final_inner(
+    state: Option<Inner<GapfillDeltaTransition>>,
+) -> Option<Vec<Option<f64>>> {
+    state.map(|mut s| s.to_vec())
+}
+
+extension_sql!(
+    r#"
+CREATE AGGREGATE @extschema@.prom_delta(
+    lowest_time TIMESTAMPTZ,
+    greatest_time TIMESTAMPTZ,
+    step BIGINT,
+    range BIGINT,
+    sample_time TIMESTAMPTZ,
+    sample_value DOUBLE PRECISION)
+(
+    sfunc=@extschema@.prom_delta_transition,
+    stype=internal,
+    finalfunc=@extschema@.prom_delta_final
+);
+"#,
+    name = "create_prom_delta_aggregate"
+);
+
+#[cfg(any(test, feature = "pg_test"))]
+#[pg_schema]
+mod tests {
+    use pgx::*;
+
+    // This somewhat hacky solution adds this SQL to the extension's schema definition when
+    // running tests.
+    // TODO: There is no isolation of this data, or namespacing.
+    extension_sql!(
+        r#"
+        CREATE TABLE gfd_test_table(t TIMESTAMPTZ, v DOUBLE PRECISION);
+        INSERT INTO gfd_test_table (t, v) VALUES
+            ('2000-01-02 15:00:00 UTC', 0),
+            ('2000-01-02 15:05:00 UTC', 50),
+            ('2000-01-02 15:10:00 UTC', 100),
+            ('2000-01-02 15:15:00 UTC', 150),
+            ('2000-01-02 15:20:00 UTC', 200),
+            ('2000-01-02 15:25:00 UTC', 200),
+            ('2000-01-02 15:30:00 UTC', 150),
+            ('2000-01-02 15:35:00 UTC', 100),
+            ('2000-01-02 15:40:00 UTC', 50),
+            ('2000-01-02 15:45:00 UTC', 0);
+    "#,
+        name = "initialize_prom_agg_test_setup"
+    );
```

Comment:
You should be able to run these commands using the SPI at test start, see eg. [this](https://github.com/timescale/timescaledb-toolkit/blob/3e9beb1935bb6d889c55fe82dbbc7c2c785266d2/extension/src/uddsketch.rs#L657-L658).
(I personally prefer to use sub-selects to create the data, [like so](https://github.com/timescale/timescaledb-toolkit/blob/3e9beb1935bb6d889c55fe82dbbc7c2c785266d2/extension/src/time_series/pipeline/aggregation.rs#L725-L737) because it sidesteps issues with table iteration order)

Reply:
Thanks! I know that I can run the setup in the SPI at start. I was trying to work around the fact that all of the tests have the same setup SQL, and that I want to use the `#[pg_test(error = "time is null")]` macro to assert failure of the function (so I can't put all of the asserts into one function).

Coming from the Java/JavaScript world, I'd rely on some setup functionality which the testing framework provides me. In Rust it doesn't seem that that exists in the built-in testing setup, and I wasn't sure if I should pull in a library for that.

Do you have any recommendations?

I have reworked this test setup a bit. Curious to hear your thoughts.

Reply:
(haven't looked through the new code yet, so these are fairly generic thoughts)
Common setup code is a perennial issues for the rust testsuite, and there isn't really a canonical solution yet. For stuff like this, I tend towards the simplest and dumbest thing: copy-paste if the tests are logically distinct and just happen to have similar setup, or a function called at the start of each test if the inputs are logically related.

In general my expectation is that most people will encounter test code when they're trying to do a bug fix: they don't really know the surrounding code, and want to be in and out of the codebase as quickly as possible. I think keeping tests cases self-contained assists with this: you only need to read a single function to get context, and it's usually clearer what the unit under test is when using dedicated setup code than when there's common code designed to be general. I often find that code designed to factor out common test tasks can be misplaced: the time spent writing the test is usually less painful than the time spent debugging a patch or tracking down a time-sensitive bug, and after a certain point common code gets complicated enough that it can obfuscate the unit-under-test making it difficult to determine what the test means, and in really bad cases causing test failures do to things unrelated to the features or, even worse, hiding test failures. That said, I know @cevian has differences of opinion with me on testing philosophy, so he may want to comment as well.

(For full clarity and future readers, in this particular case I think a common data table has some unique issues:
1. The aforementioned ordering issue: table scans in postgres return values in an arbitrary order by default, and we have seen cases where the actual order changes across versions, or even just different runs. Since it's very easy to accidentally make a test rely on the specific order of tuples returned, I find tests using such scans to be needlessly brittle (though not enough to outright ban them).
2. Concurrency issues: I'm not entirely sure how this interacts with the pgx shims, but rust tests usually run in parallel by default; accessing the same table from multiple tests _can_ result in weird nondeterminism, though I think that's less likely here where it should be read-only.

)